src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+	const_iterator end() const { return const_iterator{&_assigned, _assigned.end()}; }
+
+	/**
+	 * Assigns a new ProtoPalAttrs in a free slot, assuming there is one
+	 * Args are passed to the `ProtoPalAttrs`'s constructor
+	 */
+	template<typename... Ts>
+	auto assign(Ts &&...args) {
+		auto freeSlot = std::find_if_not(
+			_assigned.begin(), _assigned.end(),
+			[](std::optional<ProtoPalAttrs> const &slot) { return slot.has_value(); });
+
+		if (freeSlot == _assigned.end()) { // We are full, use a new slot
+			_assigned.emplace_back(std::forward<Ts>(args)...);
+		} else { // Reuse a free slot
+			(*freeSlot).emplace(std::forward<Ts>(args)...);
+		}
+		return freeSlot;
+	}
+	void remove(iterator const &iter) {
+		(*iter._iter).reset(); // This time, we want to access the `optional` itself
+	}
+	void clear() { _assigned.clear(); }
+
+	/**
+	 * Computes the "relative size" of a proto-palette on this palette
+	 */
+	double relSizeOf(ProtoPalette const &protoPal) const {
+		return std::transform_reduce(
+			protoPal.begin(), protoPal.end(), .0, std::plus<>(), [this](uint16_t color) {
+				// NOTE: The paper and the associated code disagree on this: the code has
+			    // this `1 +`, whereas the paper does not; its lack causes a division by 0
+			    // if the symbol is not found anywhere, so I'm assuming the paper is wrong.
+				return 1.
+			           / (1
+			              + std::count_if(
+							  begin(), end(), [this, &color](ProtoPalAttrs const &attrs) {
+								  ProtoPalette const &pal = _protoPals[attrs.palIndex];
+								  return std::find(pal.begin(), pal.end(), color) != pal.end();
+							  }));
+			});
+	}
+private:
+	std::unordered_set<uint16_t> &uniqueColors() const {
+		// We check for *distinct* colors by stuffing them into a `set`; this should be
+		// faster than "back-checking" on every element (O(n²))
+		//
+		// TODO: calc84maniac suggested another approach; try implementing it, see if it
+		// performs better:
+		// > So basically you make a priority queue that takes iterators into each of your sets
+		// (paired with end iterators so you'll know where to stop), and the comparator tests the
+		// values pointed to by each iterator > Then each iteration you pop from the queue,
+		// optionally add one to your count, increment the iterator and push it back into the queue
+		// if it didn't reach the end > and you do this until the priority queue is empty
+		static std::unordered_set<uint16_t> colors;
+
+		colors.clear();
+		for (ProtoPalAttrs const &attrs : *this) {
+			for (uint16_t color : _protoPals[attrs.palIndex]) {
+				colors.insert(color);
+			}
+		}
+		return colors;
+	}
+public:
+	/**
+	 * Returns the number of distinct colors
+	 */
+	size_t volume() const { return uniqueColors().size(); }
+	bool canFit(ProtoPalette const &protoPal) const {
+		auto &colors = uniqueColors();
+		for (uint16_t color : protoPal) {
+			colors.insert(color);
+		}
```

Comment:
As of C++17: (same for any other for-each-insert loops)
```suggestion
		colors.merge(protoPal);
```

Reply:
This specific case is not possible because `protoPal` is not an `unordered_set`, despite being iterable on.

Reply:
How about `colors.insert(protoPal.begin(), protoPal.end());`?

Reply:
Yup, I missed that one.